File: frontend/src/lib/bookings.ts
/**
 * Bookings API client
 * Functions to interact with the bookings backend API
 */

import { API_BASE_URL, API_ENDPOINTS } from "@/config/api";
import { TransportLocation, Booking } from "@/types";

/**
 * Fetch all active transport locations
 */
export async function getTransportLocations(): Promise<TransportLocation[]> {
  const response = await fetch(`${API_BASE_URL}/api/services/locations`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
    },
    cache: "no-store",
  });

  if (!response.ok) {
    throw new Error(`Failed to fetch transport locations: ${response.statusText}`);
  }

  const data = await response.json();
  return data;
}

/**
 * Check availability for a date range
 */
export async function checkAvailability(
  startDate: string,
  endDate?: string,
  days: number = 7
): Promise<any> {
  const params = new URLSearchParams({
    start_date: startDate,
    days: days.toString(),
  });

  if (endDate) {
    params.set("end_date", endDate);
  }

  const response = await fetch(
    `${API_BASE_URL}${API_ENDPOINTS.BOOKINGS.AVAILABILITY}?${params}`,
    {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
      cache: "no-store",
    }
  );

  if (!response.ok) {
    throw new Error(`Failed to fetch availability: ${response.statusText}`);
  }

  return response.json();
}

/**
 * Create a new booking
 */
export async function createBooking(
  bookingData: {
    package_id: string;
    booking_date: string;
    booking_time: string;
    location_id?: string;
    custom_location_address?: string;
    custom_location_latitude?: number;
    custom_location_longitude?: number;
    custom_location_distance_km?: number;
    num_brides: number;
    num_maids: number;
    num_mothers: number;
    num_others: number;
    wedding_theme?: string;
    special_requests?: string;
    guest_email?: string;
    guest_name?: string;
    guest_phone?: string;
  },
  token?: string
): Promise<Booking> {
  const headers: HeadersInit = {
    "Content-Type": "application/json",
  };

  if (token) {
    headers["Authorization"] = `Bearer ${token}`;
  }

  const response = await fetch(`${API_BASE_URL}${API_ENDPOINTS.BOOKINGS.CREATE}`, {
    method: "POST",
    headers,
    body: JSON.stringify(bookingData),
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || `Failed to create booking: ${response.statusText}`);
  }

  const data = await response.json();
  return transformBookingResponse(data);
}

/**
 * Calculate booking price
 */
export function calculateBookingPrice(
  baseBridePrice: number = 0,
  baseMaidPrice: number = 0,
  baseMotherPrice: number = 0,
  baseOtherPrice: number = 0,
  numBrides: number = 0,
  numMaids: number = 0,
  numMothers: number = 0,
  numOthers: number = 0,
  transportCost: number = 0
): { subtotal: number; transport: number; total: number; deposit: number } {
  const subtotal =
    baseBridePrice * numBrides +
    baseMaidPrice * numMaids +
    baseMotherPrice * numMothers +
    baseOtherPrice * numOthers;

  const total = subtotal + transportCost;
  const deposit = Math.round(total * 0.5);

  return {
    subtotal,
    transport: transportCost,
    total,
    deposit,
  };
}

/**
 * Transform backend booking response (snake_case) to frontend format (camelCase)
 */
function transformBookingResponse(data: any): Booking {
  return {
    id: data.id,
    bookingNumber: data.booking_number,
    userId: data.user_id,
    guestEmail: data.guest_email,
    guestName: data.guest_name,
    guestPhone: data.guest_phone,
    packageId: data.package_id,
    bookingDate: data.booking_date,
    bookingTime: data.booking_time,
    locationId: data.location_id,
    customLocationAddress: data.custom_location_address,
    customLocationLatitude: data.custom_location_latitude,
    customLocationLongitude: data.custom_location_longitude,
    customLocationDistanceKm: data.custom_location_distance_km,
    numBrides: data.num_brides,
    numMaids: data.num_maids,
    numMothers: data.num_mothers,
    numOthers: data.num_others,
    weddingTheme: data.wedding_theme,
    specialRequests: data.special_requests,
    subtotal: parseFloat(data.subtotal),
    transportCost: parseFloat(data.transport_cost),
    totalAmount: parseFloat(data.total_amount),
    depositAmount: data.deposit_amount ? parseFloat(data.deposit_amount) : undefined,
    depositPaid: data.deposit_paid,
    depositPaidAt: data.deposit_paid_at,
    status: data.status,
    adminNotes: data.admin_notes,
    createdAt: data.created_at,
    updatedAt: data.updated_at,
    package: data.package,
    location: data.location,
  };
}

/**
 * Get a specific booking by ID
 */
export async function getBookingById(bookingId: string, token?: string): Promise<Booking> {
  const headers: HeadersInit = {
    "Content-Type": "application/json",
  };

  if (token) {
    headers["Authorization"] = `Bearer ${token}`;
  }

  const response = await fetch(`${API_BASE_URL}${API_ENDPOINTS.BOOKINGS.DETAIL(bookingId)}`, {
    method: "GET",
    headers,
    cache: "no-store",
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || `Failed to fetch booking: ${response.statusText}`);
  }

  const data = await response.json();
  return transformBookingResponse(data);
}

/**
 * Format currency for display (KSh)
 */
export function formatCurrency(amount: number | undefined | null): string {
  if (amount === undefined || amount === null) return "KSh 0";
  return `KSh ${amount.toLocaleString()}`;
}

/**
 * Format date for display
 */
export function formatDate(dateStr: string | undefined | null): string {
  if (!dateStr) return "N/A";
  const date = new Date(dateStr);
  return date.toLocaleDateString("en-US", {
    weekday: "long",
    year: "numeric",
    month: "long",
    day: "numeric",
  });
}

/**
 * Format time for display
 */
export function formatTime(timeStr: string | undefined | null): string {
  if (!timeStr) return "N/A";
  const [hours, minutes] = timeStr.split(":");
  const hour = parseInt(hours);
  const ampm = hour >= 12 ? "PM" : "AM";
  const displayHour = hour % 12 || 12;
  return `${displayHour}:${minutes} ${ampm}`;
}

/**
 * Get user's booking list with pagination
 */
export async function getUserBookings(
  token: string,
  page: number = 1,
  pageSize: number = 20,
  status?: string
): Promise<{ items: Booking[]; total: number; page: number; totalPages: number }> {
  const params = new URLSearchParams({
    page: page.toString(),
    page_size: pageSize.toString(),
  });

  if (status) {
    params.set("status", status);
  }

  const response = await fetch(`${API_BASE_URL}${API_ENDPOINTS.BOOKINGS.LIST}?${params}`, {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
    cache: "no-store",
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || `Failed to fetch bookings: ${response.statusText}`);
  }

  const data = await response.json();
  return {
    items: data.items.map(transformBookingResponse),
    total: data.total,
    page: data.page,
    totalPages: data.total_pages,
  };
}

/**
 * Cancel a booking
 */
export async function cancelBooking(bookingId: string, token: string): Promise<Booking> {
  const response = await fetch(`${API_BASE_URL}${API_ENDPOINTS.BOOKINGS.CANCEL(bookingId)}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || `Failed to cancel booking: ${response.statusText}`);
  }

  const data = await response.json();
  return transformBookingResponse(data);
}

/**
 * Get status badge variant based on booking status
 */
export function getStatusBadgeVariant(
  status: string
): "default" | "secondary" | "destructive" | "outline" {
  switch (status) {
    case "confirmed":
    case "deposit_paid":
      return "default";
    case "pending":
      return "secondary";
    case "cancelled":
      return "destructive";
    case "completed":
      return "outline";
    default:
      return "secondary";
  }
}

/**
 * Format status for display
 */
export function formatStatus(status: string): string {
  return status
    .split("_")
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join(" ");
}

File: backend/app/api/routes/admin/bookings.py
"""Admin booking management API endpoints."""
from datetime import date as date_type
from typing import Optional
from uuid import UUID
import math
import csv
import io

from fastapi import APIRouter, Depends, HTTPException, Query, status
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.dependencies import get_current_admin_user
from app.models.user import User
from app.schemas.booking import BookingListResponse, BookingResponse
from app.schemas.admin_booking import (
    AdminBookingUpdate,
    AdminBookingStatusUpdate,
    AdminBookingDepositUpdate,
)
from app.services import booking_service

router = APIRouter(prefix="/admin/bookings", tags=["admin", "bookings"])


@router.get("", response_model=BookingListResponse)
def list_all_bookings(
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(20, ge=1, le=100, description="Items per page"),
    booking_status: Optional[str] = Query(None, alias="status", description="Filter by status"),
    start_date: Optional[date_type] = Query(None, alias="startDate", description="Filter by booking date >= start_date"),
    end_date: Optional[date_type] = Query(None, alias="endDate", description="Filter by booking date <= end_date"),
    location_id: Optional[UUID] = Query(None, alias="locationId", description="Filter by location ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """
    Get list of all bookings with filters (admin only).

    **Authentication required - Admin only**

    Query parameters:
    - **page**: Page number (default: 1)
    - **page_size**: Items per page (default: 20, max: 100)
    - **status**: Filter by booking status (optional)
    - **startDate**: Filter by booking date >= start_date (optional)
    - **endDate**: Filter by booking date <= end_date (optional)
    - **locationId**: Filter by location ID (optional)

    Returns:
    - List of all bookings with full details
    - Pagination metadata
    """
    if end_date and start_date and end_date < start_date:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="end_date must be after start_date"
        )

    skip = (page - 1) * page_size

    bookings, total = booking_service.get_all_bookings(
        db=db,
        skip=skip,
        limit=page_size,
        status=booking_status,
        start_date=start_date,
        end_date=end_date,
        location_id=location_id
    )

    total_pages = math.ceil(total / page_size) if total > 0 else 1

    return BookingListResponse(
        items=bookings,
        total=total,
        page=page,
        page_size=page_size,
        total_pages=total_pages
    )


@router.get("/{booking_id}", response_model=BookingResponse)
def get_booking(
    booking_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """
    Get specific booking details (admin only).

    **Authentication required - Admin only**

    Returns detailed information for any booking (bypasses ownership check).
    """
    booking = booking_service.get_booking_by_id(db, booking_id)

    if not booking:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Booking with ID {booking_id} not found"
        )

    return booking


@router.put("/{booking_id}", response_model=BookingResponse)
def update_booking(
    booking_id: UUID,
    booking_data: AdminBookingUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """
    Update a booking (admin only).

    **Authentication required - Admin only**

    Allows admin to update booking details including:
    - Date and time (with availability check)
    - Location (with pricing recalculation)
    - Participant counts (with pricing recalculation)
    - Wedding theme and special requests
    - Admin notes

    Raises:
    - 404: Booking not found
    - 400: Validation error (invalid location, unavailable slot, etc.)
    """
    try:
        booking = booking_service.admin_update_booking(
            db=db,
            booking_id=booking_id,
            booking_date=booking_data.booking_date,
            booking_time=booking_data.booking_time,
            location_id=booking_data.location_id,
            num_brides=booking_data.num_brides,
            num_maids=booking_data.num_maids,
            num_mothers=booking_data.num_mothers,
            num_others=booking_data.num_others,
            wedding_theme=booking_data.wedding_theme,
            special_requests=booking_data.special_requests,
            admin_notes=booking_data.admin_notes,
        )
        return booking
    except ValueError as e:
        error_message = str(e)
        if "not found" in error_message.lower():
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=error_message
            )
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=error_message
        )


@router.put("/{booking_id}/deposit", response_model=BookingResponse)
def mark_deposit_paid(
    booking_id: UUID,
    deposit_data: AdminBookingDepositUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """
    Mark deposit as paid/unpaid (admin only).

    **Authentication required - Admin only**

    Effects:
    - Updates deposit_paid flag
    - Sets deposit_paid_at timestamp if marked as paid
    - Updates status to "deposit_paid" if currently "pending" and marking as paid
    - Adds admin note with timestamp

    Raises:
    - 404: Booking not found
    """
    try:
        booking = booking_service.admin_mark_deposit_paid(
            db=db,
            booking_id=booking_id,
            deposit_paid=deposit_data.deposit_paid,
            admin_notes=deposit_data.admin_notes
        )
        return booking
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )


@router.put("/{booking_id}/status", response_model=BookingResponse)
def update_booking_status(
    booking_id: UUID,
    status_data: AdminBookingStatusUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """
    Update booking status (admin only).

    **Authentication required - Admin only**

    Valid statuses:
    - pending: Awaiting confirmation
    - confirmed: Confirmed booking
    - deposit_paid: Deposit received
    - completed: Service completed
    - cancelled: Booking cancelled

    Adds admin note with status change timestamp.

    Raises:
    - 404: Booking not found
    - 400: Invalid status value
    """
    try:
        booking = booking_service.admin_update_booking_status(
            db=db,
            booking_id=booking_id,
            status=status_data.status,
            admin_notes=status_data.admin_notes
        )
        return booking
    except ValueError as e:
        error_message = str(e)
        if "not found" in error_message.lower():
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=error_message
            )
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=error_message
        )


@router.delete("/{booking_id}", status_code=status.HTTP_204_NO_CONTENT)
def cancel_booking(
    booking_id: UUID,
    admin_notes: Optional[str] = Query(None, description="Admin notes for cancellation"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """
    Cancel a booking (admin only).

    **Authentication required - Admin only**

    Admin cancellation bypasses:
    - 24-hour cancellation policy
    - Ownership check
    - Completed booking restriction

    Effects:
    - Changes status to "cancelled"
    - Frees calendar slot
    - Adds cancellation note with timestamp

    Raises:
    - 404: Booking not found
    - 400: Booking already cancelled
    """
    try:
        booking_service.admin_cancel_booking(
            db=db,
            booking_id=booking_id,
            admin_notes=admin_notes
        )
        return None
    except ValueError as e:
        error_message = str(e)
        if "not found" in error_message.lower():
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=error_message
            )
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=error_message
        )


@router.get("/export/csv", response_class=StreamingResponse)
def export_bookings_csv(
    booking_status: Optional[str] = Query(None, alias="status", description="Filter by status"),
    start_date: Optional[date_type] = Query(None, alias="startDate", description="Filter by booking date >= start_date"),
    end_date: Optional[date_type] = Query(None, alias="endDate", description="Filter by booking date <= end_date"),
    location_id: Optional[UUID] = Query(None, alias="locationId", description="Filter by location ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """
    Export bookings to CSV (admin only).

    **Authentication required - Admin only**

    Exports all bookings matching the filters to a CSV file.

    Query parameters (same as list endpoint):
    - **status**: Filter by booking status (optional)
    - **startDate**: Filter by booking date >= start_date (optional)
    - **endDate**: Filter by booking date <= end_date (optional)
    - **locationId**: Filter by location ID (optional)

    Returns:
    - CSV file download with booking data
    """
    # Get all bookings (no pagination for export)
    bookings, _ = booking_service.get_all_bookings(
        db=db,
        skip=0,
        limit=10000,  # High limit for export
        status=booking_status,
        start_date=start_date,
        end_date=end_date,
        location_id=location_id
    )

    # Create CSV in memory
    output = io.StringIO()
    writer = csv.writer(output)

    # Write header
    writer.writerow([
        "Booking Number",
        "Status",
        "Date",
        "Time",
        "Customer Name",
        "Customer Email",
        "Customer Phone",
        "Package ID",
        "Location ID",
        "Num Brides",
        "Num Maids",
        "Num Mothers",
        "Num Others",
        "Wedding Theme",
        "Subtotal",
        "Transport Cost",
        "Total Amount",
        "Deposit Amount",
        "Deposit Paid",
        "Deposit Paid At",
        "Admin Notes",
        "Created At",
    ])

    # Write booking data
    for booking in bookings:
        customer_name = booking.guest_name if booking.guest_name else f"User ID: {booking.user_id}"
        customer_email = booking.guest_email if booking.guest_email else "N/A"
        customer_phone = booking.guest_phone if booking.guest_phone else "N/A"

        writer.writerow([
            booking.booking_number,
            booking.status,
            booking.booking_date,
            booking.booking_time,
            customer_name,
            customer_email,
            customer_phone,
            str(booking.package_id),
            str(booking.location_id),
            booking.num_brides,
            booking.num_maids,
            booking.num_mothers,
            booking.num_others,
            booking.wedding_theme or "",
            float(booking.subtotal),
            float(booking.transport_cost),
            float(booking.total_amount),
            float(booking.deposit_amount) if booking.deposit_amount else 0,
            booking.deposit_paid,
            booking.deposit_paid_at.isoformat() if booking.deposit_paid_at else "",
            booking.admin_notes or "",
            booking.created_at.isoformat(),
        ])

    # Prepare response
    output.seek(0)
    filename = f"bookings_export_{date_type.today().strftime('%Y%m%d')}.csv"

    return StreamingResponse(
        iter([output.getvalue()]),
        media_type="text/csv",
        headers={"Content-Disposition": f"attachment; filename={filename}"}
    )
